### ISSUE TITLE
Matplotlib recipe fails sometimes

### ISSUE THREAD

**Issue by inclement** [kivy/python-for-android#1859]
Marcello Musarra reports on the mailing list:

```[INFO]:    -> running python setup.py build_ext -v
           working: distutils.errors.DistutilsErException in thread background thread for pid 28966:
Traceback (most recent call last):
  File "/usr/lib/python2.7/threading.py", line 801, in __bootstrap_inner
    self.run()
  File "/usr/lib/python2.7/threading.py", line 754, in run
    self.__target(*self.__args, **self.__kwargs)
  File "/usr/local/lib/python2.7/dist-packages/sh-1.12.14-py2.7.egg/sh.py", line 1540, in wrap
    fn(*args, **kwargs)
  File "/usr/local/lib/python2.7/dist-packages/sh-1.12.14-py2.7.egg/sh.py", line 2459, in background_thread
    handle_exit_code(exit_code)
  File "/usr/local/lib/python2.7/dist-packages/sh-1.12.14-py2.7.egg/sh.py", line 2157, in fn
    return self.command.handle_command_exit_code(exit_code)
  File "/usr/local/lib/python2.7/dist-packages/sh-1.12.14-py2.7.egg/sh.py", line 815, in handle_command_exit_code
    raise exc
ErrorReturnCode_1:

  RAN: /home/musarra/Documents/pyprojects/uface/.buildozer/android/platform/build/build/other_builds/hostpython3/desktop/hostpython3/native-build/python setup.py build_ext -v

  STDOUT:
============================================================================
Edit setup.cfg to change the build options

BUILDING MATPLOTLIB
            matplotlib: yes [3.0.3]
                python: yes [3.7.1 (default, Jun  9 2019, 12:14:20)  [GCC
                        7.4.0]]
              platform: yes [linux]

REQUIRED DEPENDENCIES AND EXTENSIONS
                 numpy: yes [not found. pip may install it below.]
      install_requires: yes [handled by setuptools]
                libagg: yes [pkg-config information for 'libagg' could not
                        be found. Using local copy.]
              freetype: yes [version 2.8.1]
                   png: yes [version 1.6.34]
                 qhull: yes [pkg-config information for 'libqhull' could not
                        be found. Using local copy.]

OPTIONAL SUBPACKAGES
           sample_data: no  [skipping due to configuration]
              toolkits: yes [installing]
                 tests: no  [skipping due to configuration]
        toolkits_tests: no  [skipping due to configuration]

OPTIONAL BACKEND EXTENSIONS
                   agg: yes [installing]
                 tkagg: no  [Disabled by patching during Android build]
                macosx: no  [skipping due to configuration]
             windowing: no  [skipping due to configuration]

OPTIONAL PACKAGE DATA
                  dlls: no  [skipping due to configuration]

Download error on https://pypi.org/simple/numpy/: unknown url type: https -- Some packages may not be found!
Couldn't find index page for 'numpy' (maybe misspelled?)
Download error on https://pypi.org/simple/: unknown url type: https -- Some packages may not be found!
No local packages or working download links found for numpy==1.15.1
Traceback (most recent call last):
  File "setup.py", line 297, in <module>
    cmdclass=cmdclass,
  File "/home/musarra/Documents/pyprojects/uface/.buildozer/android/platform/build/build/other_builds/hostpython3/desktop/hostpython3/native-build/Lib/site-packages/setuptools/__init__.py", line 144, in setup
    _install_setup_requires(attrs)
  File "/home/musarra/Documents/pyprojects/uface/.buildozer/android/platform/build/build/other_builds/hostpython3/desktop/hostpython3/native-build/Lib/site-packages/setuptools/__init__.py", line 139, in _install_setup_requires
    dist.fetch_build_eggs(dist.setup_requires)
  File "/home/musarra/Documents/pyprojects/uface/.buildozer/android/platform/build/build/other_builds/hostpython3/desktop/hostpython3/native-build/Lib/site-packages/setuptools/dist.py", line 724, in fetch_build_eggs
    replace_conflicting=True,
  File "/home/musarra/Documents/pyprojects/uface/.buildozer/android/platform/build/build/other_builds/hostpython3/desktop/hostpython3/native-build/Lib/site-packages/pkg_resources/__init__.py", line 782, in resolve
    replace_conflicting=replace_conflicting
  File "/home/musarra/Documents/pyprojects/uface/.buildozer/android/platform/build/build/other_builds/hostpython3/desktop/hostpython3/native-build/Lib/site-packages/pkg_resources/__init__.py", line 1065, in best_match
    return self.obtain(req, installer)
  File "/home/musarra/Documents/pyprojects/uface/.buildozer/android/platform/build/build/other_builds/hostpython3/desktop/hostpython3/native-build/Lib/site-packages/pkg_resources/__init__.py", line 1077, in obtain
    return installer(requirement)
  File "/home/musarra/Documents/pyprojects/uface/.buildozer/android/platform/build/build/other_builds/hostpython3/desktop/hostpython3/native-build/Lib/site-packages/setuptools/dist.py", line 791, in fetch_build_egg
    return cmd.easy_install(req)
  File "/home/musarra/Documents/pyprojects/uface/.buildozer/android/platform/build/build/other_builds/hostpython3/desktop/hostpython3/native-build/Lib/site-packages/setuptools/command/easy_install.py", line 673, in easy_install
    raise DistutilsError(msg)
distutils.errors.DistutilsError: Could not find suitable distribution for Requirement.parse('numpy==1.15.1')


  STDERR:


Traceback (most recent call last):
  File "/usr/lib/python2.7/runpy.py", line 174, in _run_module_as_main
    "__main__", fname, loader, pkg_name)
  File "/usr/lib/python2.7/runpy.py", line 72, in _run_code
    exec code in run_globals
  File "/home/musarra/Documents/pyprojects/uface/.buildozer/android/platform/python-for-android/pythonforandroid/toolchain.py", line 1186, in <module>
    main()
  File "/home/musarra/Documents/pyprojects/uface/.buildozer/android/platform/python-for-android/pythonforandroid/toolchain.py", line 1180, in main
    ToolchainCL()
  File "/home/musarra/Documents/pyprojects/uface/.buildozer/android/platform/python-for-android/pythonforandroid/toolchain.py", line 663, in __init__
    getattr(self, args.subparser_name.replace('-', '_'))(args)
  File "/home/musarra/Documents/pyprojects/uface/.buildozer/android/platform/python-for-android/pythonforandroid/toolchain.py", line 152, in wrapper_func
    build_dist_from_args(ctx, dist, args)
  File "/home/musarra/Documents/pyprojects/uface/.buildozer/android/platform/python-for-android/pythonforandroid/toolchain.py", line 205, in build_dist_from_args
    args, "ignore_setup_py", False
  File "pythonforandroid/build.py", line 555, in build_recipes
  File "pythonforandroid/recipe.py", line 883, in build_arch
  File "pythonforandroid/recipe.py", line 938, in build_compiled_components
  File "pythonforandroid/recipe.py", line 895, in build_compiled_components
  File "pythonforandroid/logger.py", line 178, in shprint
  File "/usr/local/lib/python2.7/dist-packages/sh-1.12.14-py2.7.egg/sh.py", line 863, in next
    self.wait()
  File "/usr/local/lib/python2.7/dist-packages/sh-1.12.14-py2.7.egg/sh.py", line 792, in wait
    self.handle_command_exit_code(exit_code)
  File "/usr/local/lib/python2.7/dist-packages/sh-1.12.14-py2.7.egg/sh.py", line 815, in handle_command_exit_code
    raise exc
sh.ErrorReturnCode_1:

  RAN: /home/musarra/Documents/pyprojects/uface/.buildozer/android/platform/build/build/other_builds/hostpython3/desktop/hostpython3/native-build/python setup.py build_ext -v

  STDOUT:
============================================================================
Edit setup.cfg to change the build options

BUILDING MATPLOTLIB
            matplotlib: yes [3.0.3]
                python: yes [3.7.1 (default, Jun  9 2019, 12:14:20)  [GCC
                        7.4.0]]
              platform: yes [linux]

REQUIRED DEPENDENCIES AND EXTENSIONS
                 numpy: yes [not found. pip may install it below.]
      install_requires: yes [handled by setuptools]
                libagg: yes [pkg-config information for 'libagg' could not
                        be found. Using local copy.]
              freetype: yes [version 2.8.1]
                   png: yes [version 1.6.34]
                 qhull: yes [pkg-config information for 'libqhull' could not
                        be found. Using local copy.]

OPTIONAL SUBPACKAGES
           sample_data: no  [skipping due to configuration]
              toolkits: yes [installing]
                 tests: no  [skipping due to configuration]
        toolkits_tests: no  [skipping due to configuration]

OPTIONAL BACKEND EXTENSIONS
                   agg: yes [installing]
                 tkagg: no  [Disabled by patching during Android build]
                macosx: no  [skipping due to configuration]
             windowing: no  [skipping due to configuration]

OPTIONAL PACKAGE DATA
                  dlls: no  [skipping due to configuration]

Download error on https://pypi.org/simple/numpy/: unknown url type: https -- Some packages may not be found!
Couldn't find index page for 'numpy' (maybe misspelled?)
Download error on https://pypi.org/simple/: unknown url type: https -- Some packages may not be found!
No local packages or working download links found for numpy==1.15.1
Traceback (most recent call last):
  File "setup.py", line 297, in <module>
    cmdclass=cmdclass,
  File "/home/musarra/Documents/pyprojects/uface/.buildozer/android/platform/build/build/other_builds/hostpython3/desktop/hostpython3/native-build/Lib/site-packages/setuptools/__init__.py", line 144, in setup
    _install_setup_requires(attrs)
  File "/home/musarra/Documents/pyprojects/uface/.buildozer/android/platform/build/build/other_builds/hostpython3/desktop/hostpython3/native-build/Lib/site-packages/setuptools/__init__.py", line 139, in _install_setup_requires
    dist.fetch_build_eggs(dist.setup_requires)
  File "/home/musarra/Documents/pyprojects/uface/.buildozer/android/platform/build/build/other_builds/hostpython3/desktop/hostpython3/native-build/Lib/site-packages/setuptools/dist.py", line 724, in fetch_build_eggs
    replace_conflicting=True,
  File "/home/musarra/Documents/pyprojects/uface/.buildozer/android/platform/build/build/other_builds/hostpython3/desktop/hostpython3/native-build/Lib/site-packages/pkg_resources/__init__.py", line 782, in resolve
    replace_conflicting=replace_conflicting
  File "/home/musarra/Documents/pyprojects/uface/.buildozer/android/platform/build/build/other_builds/hostpython3/desktop/hostpython3/native-build/Lib/site-packages/pkg_resources/__init__.py", line 1065, in best_match
    return self.obtain(req, installer)
  File "/home/musarra/Documents/pyprojects/uface/.buildozer/android/platform/build/build/other_builds/hostpython3/desktop/hostpython3/native-build/Lib/site-packages/pkg_resources/__init__.py", line 1077, in obtain
    return installer(requirement)
  File "/home/musarra/Documents/pyprojects/uface/.buildozer/android/platform/build/build/other_builds/hostpython3/desktop/hostpython3/native-build/Lib/site-packages/setuptools/dist.py", line 791, in fetch_build_egg
    return cmd.easy_install(req)
  File "/home/musarra/Documents/pyprojects/uface/.buildozer/android/platform/build/build/other_builds/hostpython3/desktop/hostpython3/native-build/Lib/site-packages/setuptools/command/easy_install.py", line 673, in easy_install
    raise DistutilsError(msg)
distutils.errors.DistutilsError: Could not find suitable distribution for Requirement.parse('numpy==1.15.1')

```

I guess maybe openssl is missing, so the https download fails, but not sure why this would have worked fine for me in that case.

**Comment by AndreMiras** [comment]
TL;DR: On Ubuntu `sudo apt install libssl-dev` then clean and build again `buildozer android clean && buildozer android debug`

So that may happen if the host Python wasn't compiled with ssl support.
We can verify that in multiple ways.

1) checking the config.log
Usually in `.buildozer/android/platform/build-*/build/other_builds/hostpython3/desktop/hostpython3/native-build/config.log`
But if you don't find it there you can `find .buildozer -name config.log` and see if you see the one for `hostpython`. Checking this log you may probably see ssl support wasn't enabled for some reason e.g. missing system dependencies.

2) trying to import `_ssl` from the compiled hostpython3 binary
For instance on a working installation we should get the following:
```
.buildozer/android/platform/build-armeabi-v7a/build/other_builds/hostpython3/desktop/hostpython3/native-build/python3 -c 'import _ssl; print(_ssl)'
<module '_ssl' from '/home/andre/workspace/EtherollApp/.buildozer/android/platform/build-armeabi-v7a/build/other_builds/hostpython3/desktop/hostpython3/native-build/build/lib.linux-x86_64-3.7/_ssl.cpython-37m-x86_64-linux-gnu.so'>
```

3) actually looking for the compiled ssl module
```sh
find .buildozer/ -name "_ssl*.so"
```
On a working install we're getting at least:
```
.buildozer/android/platform/build-armeabi-v7a/build/other_builds/hostpython3/desktop/hostpython3/native-build/build/lib.linux-x86_64-3.7/_ssl.cpython-37m-x86_64-linux-gnu.so
```

**Comment by Julian-O** [comment]
Doesn't seem to be any action required. Closing as not planned.
